devito/types/tensor.py

```diff
@@ -0,0 +1,603 @@
+            kwargs["staggered"] = stagg[i] if stagg is not None else d
+            funcs.append(cls._sub_type(**kwargs))
+
+        return funcs
+
+    # Custom repr and str
+    def __str__(self):
+        st = ''.join([' %-2s,' % c for c in self])[1:-1]
+        return "Vector(%s)" % st
+
+    __repr__ = __str__
+
+    # Eval matrices functions for add/mull/evaluate
+    def _eval_at(self, var):
+
+        def entry(i):
+            return self[i]._eval_at(var[i])
+
+        comps = [entry(i) for i in range(np.max(self.shape))]
+        to = getattr(self, 'time_order', 0)
+        func = vec_func(self, self)
+        return func(name='%s%s' % (self.name, var.name),
+                    grid=self.grid, space_order=self.space_order,
+                    components=comps, time_order=to)
+
+    def _eval_add(self, other):
+        mat = [s+o for s, o in zip(self, other)]
+        func = vec_func(self, other)
+        name = "%s%s" % (self.name, getattr(other, 'name', str(other)))
+        to = getattr(self, 'time_order', 0)
+        return func(name=name, grid=self.grid, space_order=self.space_order,
+                    components=mat, time_order=to)
+
+    def _eval_scalar_mul(self, other):
+        mat = [a*other for a in self._mat]
+        func = vec_func(self, other)
+        name = "%s%s" % (self.name, getattr(other, 'name', str(other)))
+        to = getattr(self, 'time_order', 0)
+        return func(name=name, grid=self.grid, space_order=self.space_order,
+                    components=mat, time_order=to)
+
+    def _eval_scalar_rmul(self, other):
+        return self._eval_scalar_mul(other)
+
+    def _eval_transpose(self):
+        func = vec_func(self, self)
+        name = '%s_T' % self.name
+        to = getattr(self, 'time_order', 0)
+        return func(name=name, grid=self.grid, space_order=self.space_order,
+                    components=self._mat, time_order=to, transpose=not self.is_transposed)
+
+    @property
+    def div(self):
+        """
+        Divergence of the VectorFunction, creates the divergence Function.
+        """
+        return sum([getattr(self[i], 'd%s' % d.name)
+                    for i, d in enumerate(self.space_dimensions)])
+
+    @property
+    def laplace(self):
+        """
+        Laplacian of the VectorFunction, creates the Laplacian VectorFunction.
+        """
+        comps = []
+        to = getattr(self, 'time_order', 0)
+        func = vec_func(self, self)
+        comps = [sum([getattr(s, 'd%s2' % d.name) for d in self.space_dimensions])
+                 for s in self]
+        return func(name='lap_%s' % self.name, grid=self.grid,
+                    space_order=self.space_order, components=comps, time_order=to)
+
+    @property
+    def curl(self):
+        """
+        Gradient of the (3D) VectorFunction, creates the curl VectorFunction.
+        """
+        if len(self.space_dimensions) != 3:
+            error("Curl only defined in three dimensions")
```

Comment:
never use `error`. I we need to rip it off the codebase.

You need to raise an exception, otherwise the execution will proceed. 

Use the same exception and same err msg used in the free function `curl`